pkg/manager/config/config.go

```diff
@@ -17,170 +17,75 @@ package config
-}
-
-func (e *ConfigManager) initMetas() {
-	e.metas = map[mKeyType]imeta{
-		getMetaKey[config.ProxyServerOnline](): newMeta(pathPrefixProxyServer, func(cfg *config.Config) config.ProxyServerOnline {
-			return cfg.Proxy.ProxyServerOnline
-		}),
-		getMetaKey[config.LogOnline](): newMeta(pathPrefixLog, func(cfg *config.Config) config.LogOnline {
-			return cfg.Log.LogOnline
-		}),
-	}
-}
-
-func (e *ConfigManager) watchConfig(ctx context.Context, cfg *config.Config) error {
-	for _, m := range e.metas {
-		if err := func(m imeta) error {
-			_, err := e.get(ctx, m.getPrefix(), "")
-			if err != nil && errors.Is(err, ErrNoOrMultiResults) {
-				value, err := json.Marshal(m.getInitial(cfg))
-				if err != nil {
-					return err
-				}
-				if err = e.set(ctx, m.getPrefix(), "", string(value)); err != nil {
-					return err
-				}
-			}
-			e.watch(ctx, m.getPrefix(), "", func(logger *zap.Logger, evt mvccpb.Event) {
-				if obj, err := m.unmarshal(evt.Kv.Value); err != nil {
-					logger.Warn("failed unmarshal proxy config", zap.Error(err))
-					return
-				} else {
-					m.addToCh(obj)
-				}
-			})
-			return nil
-		}(m); err != nil {
-			return err
-		}
+	p, ok := cfg2Path[rf]
+	if !ok {
+		return errors.WithStack(errors.New("invalid type"))
 	}
-	return nil
-}
-
-func (e *ConfigManager) getCfg(ctx context.Context, metaKey reflect.Type) (any, error) {
-	m := e.metas[metaKey]
-	val, err := e.get(ctx, m.getPrefix(), "")
+	c, err := json.Marshal(val)
 	if err != nil {
-		return nil, err
+		return errors.WithStack(err)
 	}
-	return m.unmarshal(val.Value)
+	return e.set(ctx, pathPrefixConfig, p, c)
 }
 
-func (e *ConfigManager) setCfg(ctx context.Context, metaKey mKeyType, obj any) error {
-	m := e.metas[metaKey]
-	value, err := json.Marshal(obj)
-	if err != nil {
-		return err
+func (e *ConfigManager) GetConfig(ctx context.Context, val any) error {
+	rf := reflect.TypeOf(val)
+	if rf.Kind() == reflect.Pointer {
+		rf = rf.Elem()
+	}
+	p, ok := cfg2Path[rf]
+	if !ok {
+		return errors.WithStack(errors.New("invalid type"))
 	}
-	return e.set(ctx, m.getPrefix(), "", string(value))
-}
 
-// GetConfig queries the configuration from the config center.
-func GetConfig[T OnlineCfgTypes](ctx context.Context, e *ConfigManager, t *T) error {
-	obj, err := e.getCfg(ctx, getMetaKeyByObj(t))
+	c, err := e.get(ctx, pathPrefixConfig, p)
 	if err != nil {
 		return err
 	}
-	*t = *obj.(*T)
-	return nil
-}
-
-// SetConfig sets a configuration to the config center.
-func SetConfig[T OnlineCfgTypes](ctx context.Context, e *ConfigManager, t *T) error {
-	return e.setCfg(ctx, getMetaKeyByObj(t), t)
-}
-
-// GetCfgWatch returns the channel that contains updated configuration.
-func GetCfgWatch[T OnlineCfgTypes](e *ConfigManager) chan *T {
-	mt := e.metas[getMetaKey[T]()].(*meta[T])
-	return mt.ch
-}
 
-func (e *ConfigManager) GetProxyConfigWatch() <-chan *config.ProxyServerOnline {
-	return GetCfgWatch[config.ProxyServerOnline](e)
+	return json.Unmarshal(c.Value, val)
```

Comment:
val will be a map. Is that expected?

Reply:
Isn't val passed by the caller? He can pass `&ProxyServerOnline` or maybe a `map[string]any`.